### ISSUE TITLE
JavaParser uses too memory

### ISSUE THREAD

**Issue by corradomio** [javaparser/javaparser#3093]
I am processing (in parallel ;-) ) Elasticssearch. 
The project is composed by ~15.000 files.

The problem is that, ONLY to parse ALL files, I need MORE that 64GB ram and the 95% of the ram is allocated by the AST saved in memory/cache, and the 95% of this 95% is used by strings.

Note: there is a problem with the "'local caches"" used by JavaParser that I resolved implementing a ""custom"" version of the class that uses GLOBAL caches ;-)

There is a ""simple"" method to reduce the memory allocated, that has no impact on the current implementation: to be ensure that each (reasonable) string inside the AST is present ONLY 1 time.

It is enough to us a Map<String,String> collecting all strings: if a string is not present in the map, it is inserted as key & value, BUT if it is already present, instead to use the current string, you use the string saved in the map.

This doesn't resolved the problem at ""project"" level (the same string is present in different source files), BUT at minimum it reduce the allocated memory in each AST.

**Comment by jlerbsc** [comment]
How did you parse source files? Did you use LexicalPreservationPrinter? Can you give us more information?

**Comment by corradomio** [comment]
I implemented two classes:
"JavaParserPool"
and
"JavaParserRootsTypeSolver"

Both use the same GLOBAL cache: in this way, when a file is parsed, it not necessary to reparse it another time if it is already in cache.

The system we are implementing is able to analyze SEVERAL java projects in parallel: there are some tricks to be sure that EACH instance of JavaParserPool & JavaParserRootsTypeSolver uses the correct cache.

I also implemented another super-trick to remove the cached Typesolver inside the WeakHasMap used by Java Parser Facade, when we finish to parse & analyze a project.


However we have A LOT OF HUGE problems with JavaSymbolResolver:
1) stack overflow
2) unable to solve classes used with static methods
3) HOURS to resolve a NameExpr symbol. I have a maximum of 4 hours!!! To
resolve a SINGLE symbol.

I implemented another trick to ""ABORT"" the symbol resolution when it exceed a timeout (for now 500ms).

**Comment by corradomio** [comment]
Found!

It is "LexicalPreservingPrinter"!

**Comment by corradomio** [comment]
Just as extra information: I am using JP & JSS 3.19.1

**Comment by jlerbsc** [comment]
Thanks for this additional information. Your issue is probably due to the fact that you put CompilationUnit in cache. Why are you doing this? If you encounter other issues you are free to declare a new issue with a simplified test case allowing to reproduce the problem.

**Comment by corradomio** [comment]
We analyze the source files to create the ""dependency graph""

A) We parse each file 4 times (I know, in theory it is not necessary BUT in this way I can decide if to execute a step or not)
1) to collect the list of define classes/interfaces/annotations/enums (the
""types"")
2) to create the graph of "type dependencies" (a type extends/implements other types, a type uses other types in its implementation)
3) to collect the list of defined methods
4) to create the graph of "method dependencies" (the method implementation calls other methods)

This reduce, in theory, the number of times that I have to parse the same
file.

B) when I use the symbol resolver to resolve a symbol, it can use the type/method definition present in another source file. In this case it has to parse the file and saves it in its LOCAL cache for reusing.

This means that:
1) we have to process the same file 4 times
2) each file have to process other files one or more times
3) the AST of the same file is saved inside the local caches MULTIPLE times

Using a global cache, we can reduce the number of times that the same file is parsed. AND we are able to reduce the quantity of memory allocated because
each AST is present in memory ONLY 1 TIME.

This means that to use a GLOBAL cache is ABSOLUTELY MANDATORY.

**Comment by jlerbsc** [comment]
That's what I thought, you don't need to parse multiple files. I implemented this functionality (dependency graph) in a personal context by only parsing each class once.

Moreover this assumption is false. "when I use the symbol resolver to resolve a symbol, it can use the
type/method definition present in another source file. In this case **it has
to compile** the file and saves it in its LOCAL cache for reusing.". Neither JP nor the symbol solver compile source files. But maybe you want to talk about parsing instead of compilation. Whatever it is, it is not necessary to parse the source files several times.

**Comment by corradomio** [comment]
You are right, wrong term: PARSE,   not compile.

"JavaParserTypeSolver" PARSES other source files in "tryToSolveUncached".
This means that it can be necessary to PARSE multiple times the same file:
at minimum a single time from JP, BUT several other times by JPTS.

**Comment by corradomio** [comment]
Just as experiment, I implemented the ""simple"" method and applied it to Deeplearning4J (dl4j) and Elasticsearch (es)
Parsing ALL files and replacing all strings with an UNIQUE instance (using the described map) I obtained the following (gross) results:
There is ONLY the 4% of unique symbols (dl4j -> 4%, es -> 3%)
We need ONLY the 8% of memory (dl4j -> 8%, es -> 6%)

This means that is is possible to save A LOT of memory!

**Comment by jlerbsc** [comment]
Depending on java version you are using, maybe you can use JVM flags -XX:+UseStringDeduplication (-XX:+PrintStringDeduplicationStatistics) to achieve the same goal.

**Comment by corradomio** [comment]
May be and may be not:  you may not have permission to change the VM flags!

**Comment by jlerbsc** [comment]
I try to help you in my free time if that doesn't help i'm sorry.

**Comment by corradomio** [comment]
thanks, but in this case it is not necessary. I already added this idea in my application. I think this update is very useful for JP, independently by my specific case.

**Comment by jlerbsc** [comment]
This is an application problem independent of JP. This is more of an architectural issue within the application using JP. The memory is not expandable it must be managed. In this case, it is the application which is developed which retains the memory used by JP.

**Comment by Col-E** [comment]
> 3) HOURS to resolve a `NameExpr` symbol. I have a maximum of 4 hours!!! To
resolve a SINGLE symbol.

Just a note but the 3rd point mentioned in this issue is replicated in #3071 and a fix is provided in #3072

**Comment by jlerbsc** [comment]
It is not the same issue . This issue refers to poor memory management by the client application.

**Comment by Col-E** [comment]
I am not confident that the 3rd point is actually because of memory issues. They specifically point out in their 3rd point `NameExpr` resolution which currently has a exponentially increasing resolve time with the number of adjacent `NodeWithStmnt` types.

While their main issue is memory based this specific sub-item may not be caused by that, but by the bug referenced in the issue I linked. 

The user also confirms in that thread that this is something they are encountering. I have included a very simple test to demonstrate the issue in the PR linked. If you revert the changes but still run the tesf you will see the issue with _"taking forever to resolve a `NameExpr`"_ as this user mentions. The fix is quite a simple one too.
